### PR TITLE
Graphics Support for ArrayFire

### LINKED ISSUE #10: OpenCL + OpenGL interop.
Implement hooks that enable opengl inter operation in OpenCL backend.

### LINKED ISSUE #11: CUDA + OpenGL interop
Implement hooks that enable opengl inter operation in CUDA backend.

### LINKED ISSUE #549: Exception with OpenCL
Hello all

Running the simplest programs with afOpenCL.lib I get exception when the program is ended. No problem when using afCPU.lib.

The most basic program is:
``` C++
    try {
		setDevice(0);
		info();
		array a = randu(4,5,2);
    } catch (af::exception& e) {
		printf("Error: %s\n", e.what());
    }
```
Output is:
```
ArrayFire v3.0.beta (OpenCL, 64-bit Windows, build a4a26ce)
[0] INTEL   : Intel(R) HD Graphics 4000
-1- INTEL   : Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz
-2- Experimental OpenCL 2.0 CPU Only Platform: Intel(R) Core(TM) i5-3320M CPU @
2.60GHz
```
Exception is:
```
Unhandled exception at 0x000007FED84CBDA1 (igdrcl64.dll) in helloworld_debug.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.
```
Running full helloworld.cpp demo, exception is slightly different:
```
Unhandled exception at 0x0000000076CBF279 (ntdll.dll) in helloworld_debug.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.
```

### PR DESCRIPTION
**Pull request by 9prady9**
Fixes #11 
Fixes #10 
Fixes #549 
Related to #412
Added the following image processing examples
* edge  detectors
* morphological operations

### CODE REVIEW

File: src/api/c/device.cpp [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 #include <af/dim4.hpp>
 #include <af/device.h>
+#include <graphics_common.hpp>
```
Comment (pavanky) [inline src/api/c/device.cpp]: Is this necessary ?

File: src/api/c/graphics_common.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+INSTANTIATE_GET_GL_TYPE(unsigned, GL_UNSIGNED_INT);
+INSTANTIATE_GET_GL_TYPE(char, GL_BYTE);
+INSTANTIATE_GET_GL_TYPE(unsigned char, GL_UNSIGNED_BYTE);
+
+GLenum glErrorSkip(const char *msg, const char* file, int line)
+{
+#ifndef NDEBUG
+    GLenum x = glGetError();
+    if (x != GL_NO_ERROR) {
+        printf("GL Error Skipped at: %s:%d Message: %s Error Code: %d \"%s\"\n", file, line, msg, x, gluErrorString(x));
```
Comment (pavanky) [inline src/api/c/graphics_common.cpp]: This should be AF_ERROR(MSG, AF_ERR_INTERNAL);

File: src/api/c/graphics_common.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+GLenum glErrorCheck(const char *msg, const char* file, int line)
+{
+// Skipped in release mode
+#ifndef NDEBUG
+    GLenum x = glGetError();
+
+    if (x != GL_NO_ERROR) {
+        printf("GL Error at: %s:%d Message: %s Error Code: %d \"%s\"\n", file, line, msg, x, gluErrorString(x));
```
Comment (pavanky) [inline src/api/c/graphics_common.cpp]: Should be AF_ERROR

File: src/api/c/graphics_common.cpp [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    return 0;
+#endif
+}
+
+GLenum glForceErrorCheck(const char *msg, const char* file, int line)
+{
+    GLenum x = glGetError();
+
+    if (x != GL_NO_ERROR) {
+        printf("GL Error at: %s:%d Message: %s Error Code: %d \"%s\"\n", file, line, msg, x, gluErrorString(x));
```
Comment (pavanky) [inline src/api/c/graphics_common.cpp]: Add this to AF_ERROR below

File: examples/graphics/morphing.cpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    array th = tophat(img_rgb, mask);
+    array bh = bottomhat(img_rgb, mask);
+    array bl = blur(img_rgb, gaussiankernel(5,5));
+    array bp = border(img_rgb, 20, 30, 40, 50, 0.5);
+    array bo = border(img_rgb, 20);
+
+    while (!wnd.close()) {
+        wnd.grid(3, 4);
+
+        wnd(0, 0)->image(img_rgb, "Input"          );
```
Comment (umar456) [inline examples/graphics/morphing.cpp]: Window operator() should return reference. This syntax looks weird.

File: src/api/c/hist.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    hist->setYAxisTitle("Frequency");
+
+    copy_histogram<T>(histogramInput, hist);
+
+    return hist;
+}
+#endif
+
+af_err af_draw_hist(const af_window wind, const af_array X, const double minval, const double maxval,
+                   const af_cell* const props)
```
Comment (pavanky) [inline src/api/c/hist.cpp]: indent properly

File: examples/graphics/morphing.cpp (hunk 5 above)
Comment (pavanky) [inline examples/graphics/morphing.cpp]: This is large change. Fix it post merge.

File: include/af/graphics.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+typedef unsigned long long af_window;
+
+typedef struct {
+    int row;
+    int col;
+    const char* title;
+} af_cell;
+
+#ifdef __cplusplus
+#include <utility>
```
Comment (umar456) [inline include/af/graphics.h]: is this necessary?

File: include/af/graphics.h [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+{
+
+    // FIXME handle copying properly
+class AFAPI Window {
+    private:
+        af_window wnd;
+        /* below attributes are used to track which
+         * cell in the grid is being rendered currently */
+        int _r;
+        int _c;
```
Comment (umar456) [inline include/af/graphics.h]: can these values be part of af_window?

File: include/arrayfire.h [hunk 9]
```diff
@@ -230,6 +230,7 @@
 #include "af/exception.h"
 #include "af/features.h"
 #include "af/gfor.h"
+#include "af/graphics.h"
```
Comment (umar456) [inline include/arrayfire.h]: I think this should be included separately.

File: src/api/c/device.cpp (hunk 1 above)
Comment (9prady9) [inline src/api/c/device.cpp]: @pavanky probably an residue, will remove it.

File: src/api/c/graphics_common.cpp (hunk 2 above)
Comment (9prady9) [inline src/api/c/graphics_common.cpp]: Taking care of this.

File: src/api/c/graphics_common.cpp (hunk 3 above)
Comment (9prady9) [inline src/api/c/graphics_common.cpp]: I didn't quite understand what you refer to here. Should i add AF_ERROR_GL_ERR to af_err enum ?

File: examples/graphics/morphing.cpp (hunk 5 above)
Comment (9prady9) [inline examples/graphics/morphing.cpp]: Not really a big change in API, examples will change more thought. Let me do it now itself. Another PR for a minor change is not necessary.

File: include/af/graphics.h (hunk 7 above)
Comment (9prady9) [inline include/af/graphics.h]: Yes, what will C-API users do ? without such structure, the API will keep changing if we add more options to each cell rendering in multi-view mode.

Comment (pavanky) [inline include/af/graphics.h]: Not the struct, the header file.

File: include/af/graphics.h (hunk 8 above)
Comment (9prady9) [inline include/af/graphics.h]: They are part of af_window. They are required here as well, how else will a call the following will work.
`Window(1,1).draw(image);`

Comment (pavanky) [inline include/af/graphics.h]: af_window comes from forge. SO probably no.

File: include/arrayfire.h (hunk 9 above)
Comment (9prady9) [inline include/arrayfire.h]: Do you mean the user has to do it himself ?

Comment (pavanky) [inline include/arrayfire.h]: I think this is fine because we are compiling the functions even if they are not supported.

File: CMakeLists.txt [hunk 10]
```diff
@@ -35,6 +37,40 @@ ELSE(FREEIMAGE_FOUND)
...
+    IF(USE_SYSTEM_FORGE)
+        FIND_PACKAGE(Forge)
+    ELSE(USE_SYSTEM_FORGE)
+        INCLUDE("${CMAKE_MODULE_PATH}/build_forge.cmake")
+    ENDIF(USE_SYSTEM_FORGE)
+
+    IF(FORGE_FOUND)
+        ADD_DEFINITIONS(-DGLEW_MX -DWITH_GRAPHICS)
+        INCLUDE("${CMAKE_MODULE_PATH}/FindGLEWmx.cmake")
+        FIND_PACKAGE(GLFW)
```
Comment (pavanky) [inline CMakeLists.txt]: This should go away.

File: src/api/c/hist.cpp (hunk 6 above)
Comment (9prady9) [inline src/api/c/hist.cpp]: Taken care of this.

File: src/api/c/graphics_common.cpp (hunk 3 above)
Comment (9prady9) [inline src/api/c/graphics_common.cpp]: Done.

File: src/api/c/graphics_common.cpp (hunk 4 above)
Comment (9prady9) [inline src/api/c/graphics_common.cpp]: Took care of it.

File: include/af/graphics.h (hunk 7 above)
Comment (9prady9) [inline include/af/graphics.h]: Ah ok, took care of it.

File: src/backend/opencl/program.cpp [hunk 11]
```diff
@@ -6,13 +6,12 @@
...
-#include <iostream>
-#include <cl.hpp>
-#include <traits.hpp>
+
 #include <program.hpp>
+#include <traits.hpp>
 #include <kernel_headers/KParam.hpp>
-#include <platform.hpp>
 #include <debug_opencl.hpp>
+#include <iostream>
```
Comment (umar456) [inline src/backend/opencl/program.cpp]: I don't think this is required

**Comment by 9prady9** [comment]
This is ready for merge, wait for Jenkins builds to finish.

File: src/backend/opencl/program.cpp (hunk 11 above)
Comment (9prady9) [inline src/backend/opencl/program.cpp]: @umar456 we can fix this in devel later as hot fix.

**Pull request merged**